### ISSUE TITLE
Poor performance of plugins using the Hirshberg Algorithm

### ISSUE THREAD

**Issue by idawson-gl** [zaproxy/zaproxy#6515]
**Describe the bug**
Three plugins that use the [Hirshberg](https://en.wikipedia.org/wiki/Hirschberg%27s_algorithm) algorithm have been observed to use significant processor time. 
- LdapInjectionScanRule
- SourceCodeDisclosureFileInclusionScanRule
- UsernameEnumerationScanRule

A scan that took 25 minutes to complete took only 1m 25s after disabling those plugins.

These plugins can take 10 minutes to analyze only 16~40 requests. No other plugins take this amount of time to execute.

**To Reproduce**
1. Configure and run an active scan
2. Watch the active scan plugin statistics, or attach a profiler. 

**Expected behavior**
Faster runtimes for scans

**Screenshots**

Plugin stats output:

![image](https://user-images.githubusercontent.com/60455448/113529006-3d248480-95fd-11eb-9aaa-18a99b2bfdd1.png)

Profiler of an active scan:
![image](https://user-images.githubusercontent.com/60455448/113529011-40b80b80-95fd-11eb-8986-5635c1230a2e.png)

![image](https://user-images.githubusercontent.com/60455448/113529053-5e857080-95fd-11eb-8229-7cb922f28525.png)

**Software versions**
 - ZAP: 2.10.0
 - Add-on: See bug description
 - OS: Ubuntu Linux
 - Java: 11.0.10

**Errors from the zap.log file**
See https://www.zaproxy.org/faq/what-is-the-default-directory-that-zap-uses/ for the location of the zap.log file for your OS.

**Additional context**
The Hirshberg Algorithm uses the levenshtein distance algorithm and is known to be a poor fit for doing differential analysis against large bodies of text. It's time complexity is O(n*m), and apparently at best O(n +d^2) https://onak.pl/download/publications/Andoni_Krauthgamer_Onak-edit_distance.pdf.

**Would you like to help fix this issue?**
Just say if you'd like to try and help fix this issue. We'll provide any advice and guidance you need.

**Comment by kingthorin** [comment]
Yup, unfortunately this is a known weakness of these scan rules. In fact if I recall correctly it is part (or most) of the reason the LDAP rule was demoted.

We’re open to more performant suggestions with regard to comparing the similarity of responses. If anyone has ideas.

**Comment by thc202** [comment]
It was the reason: https://github.com/zaproxy/zap-extensions/blob/master/addOns/ascanrulesBeta/CHANGELOG.md#18---2015-12-04

**Comment by pranavsaxena17** [comment]
Some algorithms that could probably be used here are Jaccard similarity, Dice’s coefficient ,Cosine similarity , Manhattan distance , and Euclidean distance. I'm investigating the time taken by each of these, would update soon with results.

**Comment by ricekot** [comment]
Since this involves comparing two large bodies of text, maybe you could also take a look at the algorithms used by `git`? You can probably also find a library that already implements them (e.g. JGit).
- https://link.springer.com/article/10.1007/s10664-019-09772-z
- https://blog.jcoglan.com/2017/02/12/the-myers-diff-algorithm-part-1/

**Comment by pranavsaxena17** [comment]
I ran a benchmark for the algos,which can be found [here](https://pastebin.com/dq9qQShv)(urls to get the responses can be added at line#88 ). The code isn't pretty and can definitely be further optimized, but for the purpose of deciding between the various algorithms it should do. Upon executing, we get (note that to get further accurate results, one can run each of these seperately to account for JVM optimisations between runs of the algos ) :-

Note:- Same Response means an identical response, Similar Response means 25% of the words changed while Different Response means 75% of the words changed


The results for the various similarities, in the order of Same Response, Similar Response and Different Response are:-

Cosine Similarity

Similarity :0.9932222729152504
Time Elapsed(s) : 0.433

Similarity :0.9642797502234015
Time Elapsed(s) : 0.436

Similarity :0.5655515944937145
Time Elapsed(s) : 0.447



Jaccard Similarity

Similarity :1.0
Time Elapsed(s) : 0.012

Similarity :0.989247311827957
Time Elapsed(s) : 0.008

Similarity :0.9247311827956989
Time Elapsed(s) : 0.009



Dice Similarity

Similarity :1.0
Time Elapsed(s) : 0.023

Similarity :0.6978652097261254
Time Elapsed(s) : 0.016

Similarity :0.27792139062468196
Time Elapsed(s) : 0.018



Euclid Similarity

Similarity :1.0
Time Elapsed(s) : 0.417
Similarity :0.8685315637041576
Time Elapsed(s) : 0.416
Similarity :0.175472145707666
Time Elapsed(s) : 0.433


From the above results, and considering the fact that Jaccard seems to perform very poorly on sites such as https://nodejs.org/dist/latest-v14.x/docs/api/n-api.html ,probably because they contain a lot of occurences of a word(napi in this case), it seems that the **Dice Similarity** could be the way forward here. It's fast, roughly takes half the space of the original response and seems to give accurate measurements of  the similarity.

For further references, one could have a look at:-
- https://core.ac.uk/download/pdf/295626466.pdf (Section 3.1.1)
- https://programminghistorian.org/en/lessons/common-similarity-measures#how-to-know-which-distance-measure-to-use (everything before the 'Calculating Distance in Python' section is worth giving a read)
- http://www2.imm.dtu.dk/pubdb/edoc/imm6364.pdf (quite a long and thorough text, but most of it is not relevant to our use case. A read of sections corresponding to the algorithms used here and Chapters 6,7,8 and 9 should suffice)

Your inputs on how we should proceed from here would be really helpful @psiinon @thc202 @kingthorin @thc202

**Comment by psiinon** [comment]
This is great research!
Initial thoughts - how about we introduce a common generic class for making these sort of comparisons, which lives in the commonlib add-on?
All add-ons which need this functionality would use the generic class which would mean we could improve or swap algorithms as necessary by just changing the common lib.
Initially we could just include the Dice Similarity as the default (and maybe Hirshberg as an option?) and just switch via a config file option?

**Comment by thc202** [comment]
I'm +1 for the class that allows to know if two responses/strings are the equal, similar, or different (assuming all scan rules fall into those checks), but don't think we need to provide `Hirshberg` as an option, nor allowing to switch, we'd be introducing the same problem again.
